### PR TITLE
added testBookCategory

### PR DESCRIPTION
**Pull request by RCherniavskiy**

### CODE REVIEW

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        BookDto[] actual = objectMapper.readValue(result.getResponse().getContentAsByteArray(), BookDto[].class);
+        Assertions.assertEquals(3, actual.length);
+        for (BookDto book : expected) {
+            Assertions.assertTrue(Arrays.stream(actual)
+                    .anyMatch(a -> a.getId().equals(book.getId())
+                            && a.getTitle().equals(book.getTitle())
+                            && a.getAuthor().equals(book.getAuthor())
+                            && a.getPrice().equals(book.getPrice())
+                            && a.getIsbn().equals(book.getIsbn())));
+        }
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: containsAll() should work here and will be much shorter (if you have proper equals of course)

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto actualBook = objectMapper.readValue(result.getResponse().getContentAsByteArray(), BookDto.class);
+
+        Assertions.assertEquals(expectedBook.getId(), actualBook.getId());
+        Assertions.assertEquals(expectedBook.getTitle(), actualBook.getTitle());
+        Assertions.assertEquals(expectedBook.getAuthor(), actualBook.getAuthor());
+        Assertions.assertEquals(expectedBook.getPrice(), actualBook.getPrice());
+        Assertions.assertEquals(expectedBook.getIsbn(), actualBook.getIsbn());
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: EqualsBuilder or implementing equals in model will allow to get rid of each field comparison

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                .setDescription("Book Description")
+                .setIsbn("9460306332081")
+                .setPrice(BigDecimal.valueOf(10.99));
+
+        BookDto expected = new BookDto()
+                .setAuthor(requestDto.getAuthor())
+                .setTitle(requestDto.getTitle())
+                .setDescription(requestDto.getDescription())
+                .setIsbn(requestDto.getIsbn())
+                .setPrice(requestDto.getPrice());
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: what is sense of having 2 identical dtos? you may omit expected and use requestDto as reference

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                    .content(jsonRequest)
+                    .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isOk())
+                .andReturn();
+        BookDto actual = objectMapper.readValue(result.getResponse().getContentAsString(), BookDto.class);
+        Assertions.assertNotNull(actual);
+        Assertions.assertNotNull(actual.getId());
+        Assertions.assertEquals(expected.getAuthor(), actual.getAuthor());
+        EqualsBuilder.reflectionEquals(expected,actual,"id");
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: why check author separately if you check it during reflectionEquals?

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                book.setTitle("Book Title");
+                book.setDescription("Book Description");
+                book.setIsbn("9460306332081");
+                book.setPrice(BigDecimal.valueOf(10.99));
+        mockMvc.perform(
+                        delete("/books/{id}", 4L)
+                                .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isNoContent());
+    }
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: Why you created book? You don't use it further.
Also would be nice to recheck DB state, if book was really deleted.
If your controller has code "return new ResponseEntity(NO_CONTENT);" this test will pass

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                .setPrice(BigDecimal.valueOf(15.99));
+
+        String jsonRequest = objectMapper.writeValueAsString(updateRequestDto);
+
+        mockMvc.perform(
+                        put("/books/{id}", 1L)
+                                .content(jsonRequest)
+                                .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isOk());
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: again, nobody knows after this test if book was updated or not. Only response status is checked what means nearly nothing

File: src/test/java/book_store/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+    private static void teardown(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            Statement statement = connection.createStatement();
+            statement.executeUpdate("DELETE FROM books");
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
Comment (olekskov) [inline src/test/java/book_store/controller/BookControllerTest.java]: empty line in the end of file

File: src/test/java/book_store/controller/CategoryControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+    private static void teardown(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            Statement statement = connection.createStatement();
+            statement.executeUpdate("DELETE FROM books");
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
Comment (olekskov) [inline src/test/java/book_store/controller/CategoryControllerTest.java]: empty line

File: src/test/java/book_store/service/BookServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        Category category = new Category();
+        category.setId(categoryId);
+
+        when(bookRepository.findAllByCategoriesContaining(any(Category.class))).thenReturn(Collections.emptyList());
+
+        bookService.getBooksByCategoryId(categoryId);
+
+        verify(bookRepository, times(1)).findAllByCategoriesContaining(any(Category.class));
+    }
+}
```
Comment (olekskov) [inline src/test/java/book_store/service/BookServiceImplTest.java]: again

File: src/test/java/book_store/service/CategoryServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        verify(categoryMapper, times(1)).toDto(categoryToUpdate);
+    }
+
+    @Test
+    void deleteById_ExistingCategoryId_DeletesCategory() {
+        long categoryId = 1L;
+        categoryService.deleteById(categoryId);
+        verify(categoryRepository, times(1)).deleteById(categoryId);
+    }
+}
```
Comment (olekskov) [inline src/test/java/book_store/service/CategoryServiceImplTest.java]: recheck all files please

**Review by olekskov — CHANGES_REQUESTED** [review]

**Review by Rommelua — APPROVED** [review]

**Pull request merged**